src/netaddress.h

```diff
@@ -49,18 +51,50 @@ enum Network
     NET_MAX,
 };
 
+/// If an IPv6 address begins with this, then we treat the rest of it as IPv4 address.
+static constexpr uint8_t IPv4_IN_IPv6_PREFIX[12] = {
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF
+};
+
+/// If an IPv6 address begins with this, then we treat the rest of it as TORv2 address.
+static constexpr uint8_t TORv2_IN_IPv6_PREFIX[6] = {
+    0xFD, 0x87, 0xD8, 0x7E, 0xEB, 0x43
+};
+
+/// If an IPv6 address begins with this, then we treat the rest of it as an internal address.
+static constexpr uint8_t INTERNAL_IN_IPv6_PREFIX[6] = {
+    0xFD, 0x6B, 0x88, 0xC0, 0x87, 0x24 // 0xFD + sha256("bitcoin")[0:5].
+};
+
+/// Size of IPv4 address (in bytes).
+static constexpr size_t ADDR_IPv4_SIZE = 4;
+
+/// Size of IPv6 address (in bytes).
+static constexpr size_t ADDR_IPv6_SIZE = 16;
+
+/// Size of TORv2 address (in bytes).
+static constexpr size_t ADDR_TORv2_SIZE = 10;
+
+/// Size of "internal" (NET_INTERNAL) address (in bytes).
+static constexpr size_t ADDR_INTERNAL_SIZE = 10;
+
 /**
  * Network address.
  */
 class CNetAddr
 {
     protected:
+        /**
+         * Raw representation of the network address.
+         * In network byte order (big endian) for IPv4 and IPv6.
+         */
+        prevector<ADDR_IPv6_SIZE, uint8_t> m_addr{ADDR_IPv6_SIZE, 0x0};
```

Comment:
Are you sure the use of `prevector` is warranted here? Measurements would be nice.

Generally I think we should try stick with to the standard containers unless we have clear quantitative evidence showing that they are unsuitable for the scenario being considered. Reviewers, static analysers and compilers tend to be better at analysing use of the standard containers.

FWIW: 

```
$ git grep -l prevector -- "*.cpp" "*.h" ":(exclude)src/bench/" ":(exclude)src/test/"
src/hash.h
src/memusage.h
src/prevector.h
src/script/script.h
src/serialize.h
```

Reply:
I suggested this, so let me back it up.

Currently an address in CAddrMan consumes 256 bytes (cost in mapInfo and mapAddr together, plus allocation overhead).

This PR as-is raises that to 288 bytes, for addresses that take up to 16 bytes (which is all of them, for now).

Using an `std::vector<uint8_t>` instead would change that to 352 bytes per address, or an extra 64 on top. It would also increase memory fragmentation and (slightly) increase CPU usage because of two malloc's per address.

FTR, my node has 69691 addresses in CAddrMan.

Reply:
@sipa Thanks for quantifying! That answers my question :)

Reply:
Thanks for the question and the elaborate explanation! Indeed smaller size and avoiding an extra heap allocation is better. FWIW I moved `m_addr` definition before `m_net` to save 8 bytes from the size of `CNetAddr` (32 vs 40 bytes on x64).